pkg/rpcmiddleware/rpcretry/client_interceptors.go

```diff
@@ -0,0 +1,125 @@
+// Copyright © 2022 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rpcretry
+
+import (
+	"context"
+	"strconv"
+	"time"
+
+	"github.com/grpc-ecosystem/go-grpc-middleware/util/backoffutils"
+	"go.thethings.network/lorawan-stack/v3/pkg/log"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+type attemptMetadataKeyType struct{}
+
+var attemptMetadataKey attemptMetadataKeyType
```

Comment:
Was originally intended as a bit of extra information for outgoing context, that way in the future if need be services could take that bit of information if need be. Since I considered mostly useful for the stream interceptor is something that I should have removed, will do so